### PR TITLE
Construct path to the version file from the current directory and fil…

### LINKED ISSUE #22450: [8.2.2] Cron job creates error
I obtain the following erros every time the cron job runs (15 Minutes)

 - for each user the following error:

```
{"reqId":"p1V8PVOKpDOUNU8teqUv","remoteAddr":"","app":"PHP","message":"opendir(\/mnt\/owncloud\/data\/USERNAME\/files_trashbin\/files): failed to open dir: No such file or directory at \/var\/www\/html\/owncloud\/lib\/private\/files\/storage\/local.php#96","level":3,"time":"2016-02-16T18:30:04+01:00","method":"--","url":"--"}
```
 - then another error comes:
```
{"reqId":"p1V8PVOKpDOUNU8teqUv","remoteAddr":"","app":"PHP","message":"filesize(): stat failed for \/mnt\/owncloud\/data\/carlos\/files_versions\/rsions\/hello carlos.txt.v1444125868 at \/var\/www\/html\/owncloud\/lib\/private\/files\/storage\/local.php#139","level":3,"time":"2016-02-16T18:30:05+01:00","method":"--","url":"--"}
```
This file should be already removed, because the retention oblation is older than 3 months:
```
-rw-r--r--  1 www-data www-data   117 Okt  6 12:16 hello carlos.txt.v1444125868
```

and the path has an error: ```\/rsions\/```

Server stack:

• Operating system (name and version number):
```
Ubuntu 14.04.3 LTS
```

• URL from your ownCloud
https://cloud.damken.com/

• Your ownCloud availability:
over internet

• Webserver (name and version number):
```
Server version: Apache/2.4.7 (Ubuntu)
Server built:   Oct 14 2015 14:20:21
```

• Apache modules enabled as declared in:
```
access_compat.load
auth_basic.load
authz_core.load
autoindex.conf
deflate.load
env.load
mime.conf
mpm_prefork.load
php5.conf
setenvif.conf
ssl.conf
status.load
alias.conf
authn_core.load
authz_host.load
autoindex.load
dir.conf
filter.load
mime.load
negotiation.conf
php5.load
setenvif.load
ssl.load
alias.load
authn_file.load
authz_user.load
deflate.conf
dir.load
headers.load
mpm_prefork.conf
negotiation.load
rewrite.load
socache_shmcb.load
status.conf
```

• Database (name and version number):
```
mysql  Ver 15.1 Distrib 5.5.46-MariaDB, for debian-linux-gnu (x86_64) using readline 5.2

Connection id:          578875
Current database:
Current user:           root@localhost
SSL:                    Not in use
Current pager:          stdout
Using outfile:          ''
Using delimiter:        ;
Server:                 MariaDB
Server version:         5.5.46-MariaDB-1ubuntu0.14.04.2 (Ubuntu)
Protocol version:       10
Connection:             Localhost via UNIX socket
Server characterset:    latin1
Db     characterset:    latin1
Client characterset:    utf8
Conn.  characterset:    utf8
UNIX socket:            /var/run/mysqld/mysqld.sock
Uptime:                 22 days 7 hours 3 min 20 sec

Threads: 2  Questions: 176832246  Slow queries: 1  Opens: 159  Flush tables: 2  Open tables: 122  Queries per second avg: 91.803
--------------
```

• PHP (version number):
```
PHP 5.5.9-1ubuntu4.14 (cli) (built: Oct 28 2015 01:34:46)
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.5.0, Copyright (c) 1998-2014 Zend Technologies
    with Zend OPcache v7.0.3, Copyright (c) 1999-2014, by Zend Technologies
```

• PHP cache manager
```
Package: redis-server
Status: install ok installed
Priority: optional
Section: database
Installed-Size: 744
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Architecture: amd64
Source: redis
Version: 2:2.8.4-2
Depends: libc6 (>= 2.14), libjemalloc1 (>= 2.1.1), redis-tools (= 2:2.8.4-2), adduser
Conffiles:
 /etc/default/redis-server 45e997df22f4c5c2b5c9ae159775a089
 /etc/redis/redis.conf e79ac99a5236e6ab3faf86bdd99affaf
 /etc/redis/sentinel.conf a95ae9b0242a03c5b4dd1af934478895
 /etc/init.d/redis-server 387b03762e0cf0db194e961104137b5a
 /etc/logrotate.d/redis-server 06755b99ef70d62a56cff94cbfc36de7
Description: Persistent key-value database with network interface
 Redis is a key-value database in a similar vein to memcache but the dataset
 is non-volatile. Redis additionally provides native support for atomically
 manipulating and querying data structures such as lists and sets.
 .
 The dataset is stored entirely in memory and periodically flushed to disk.
Original-Maintainer: Chris Lamb <lamby@debian.org>
Homepage: http://redis.io/
```

• PHP installed packages
```
libapache2-mod-php5 -> 5.5.9+dfsg-1ubuntu4.14
php-gettext -> 1.0.11-1
php-pear -> 5.5.9+dfsg-1ubuntu4.14
php5 -> 5.5.9+dfsg-1ubuntu4.14
php5-apcu -> 4.0.7-1build1~ubuntu14.04.1
php5-cgi -> 5.5.9+dfsg-1ubuntu4.14
php5-cli -> 5.5.9+dfsg-1ubuntu4.14
php5-common -> 5.5.9+dfsg-1ubuntu4.14
php5-curl -> 5.5.9+dfsg-1ubuntu4.14
php5-dev -> 5.5.9+dfsg-1ubuntu4.14
php5-gd -> 5.5.9+dfsg-1ubuntu4.14
php5-imagick -> 3.1.2-1build1
php5-intl -> 5.5.9+dfsg-1ubuntu4.14
php5-json -> 1.3.2-2build1
php5-libsmbclient -> 0.5.0-10.1
php5-mcrypt -> 5.4.6-0ubuntu5
php5-mysql -> 5.5.9+dfsg-1ubuntu4.14
php5-readline -> 5.5.9+dfsg-1ubuntu4.14
phpmyadmin -> 4:4.0.10-1
pkg-php-tools -> 1.11
```

• ownCloud Enterprise Server Version
```
{"installed":true,"maintenance":false,"version":"8.2.2.2","versionstring":"8.2.2","edition":""}
```

• Did you install your ownCloud with TAR files or packages?
TAR

• Tell us your migration path of the ownCloud software
8.1.0.8 -> 8.1.3.0 -> 8.1.4 -> 8.2.2

• Do you have records from your migration steps?
No

• List of owncloud-apps enabled/ in use:
```
Enabled:
  - activity: 2.1.3
  - bookmarks: 0.6
  - calendar: 0.8.1
  - contacts: 0.5.0.0
  - delete_confirmation: 1.1.0
  - direct_menu: 0.5
  - documents: 0.11.0
  - files: 1.2.0
  - files_external: 0.3.0
  - files_mv: 0.8.1a
  - files_pdfviewer: 0.7
  - files_sharing: 0.7.0
  - files_texteditor: 2.0
  - files_trashbin: 0.7.0
  - files_versions: 1.1.0
  - firstrunwizard: 1.1
  - gallery: 14.2.0
  - notifications: 0.1.0
  - ownnote: 1.05
  - provisioning_api: 0.3.0
  - registration: 0.1.0
  - sharelinks: 1.0.1
  - tasks: 0.8.1
  - templateeditor: 0.1
  - updater: 0.6
```

• The content of your config/config.php:
```
<?php
$CONFIG = array (
  'instanceid' => 'XXXXXX',
  'passwordsalt' => 'XXXXXX',
  'secret' => 'XXXXXX',
  'trusted_domains' =>
  array (
    0 => 'cloud.damken.com',
  ),
  'datadirectory' => '/mnt/owncloud/data',
  'dbtype' => 'mysql',
  'version' => '8.2.2.2',
  'dbname' => 'XXXXXX',
  'dbhost' => '127.0.0.1',
  'dbtableprefix' => 'oc_',
  'dbuser' => 'XXXXXX',
  'dbpassword' => 'XXXXXX',
  'loglevel' => 0,
  'logtimezone' => 'Europe/Berlin',
  'log_rotate_size' => 10485760,
  'installed' => true,
  'mail_smtpmode' => 'smtp',
  'mail_from_address' => 'XXXXXX',
  'mail_domain' => 'damken.com',
  'mail_smtpsecure' => 'XXXXXX',
  'mail_smtpauth' => 1,
  'mail_smtpauthtype' => 'XXXXXX',
  'mail_smtphost' => 'damken.com',
  'mail_smtpport' => 'XXXXXX',
  'mail_smtpname' => 'XXXXXX',
  'mail_smtppassword' => 'XXXXXX',
  'theme' => 'damken',
  'maintenance' => false,
  'appstoreenabled' => true,
  'appstoreurl' => 'https://api.owncloud.com/v1',
  'appstore.experimental.enabled' => true,
  'memcache.local' => '\\OC\\Memcache\\ACPu',
  'filelocking.enabled' => 'true',
  'memcache.distributed' => '\\OC\\Memcache\\Redis',
  'memcache.locking' => '\\OC\\Memcache\\Redis',
  'redis' =>
  array (
    'host' => '127.0.0.1',
    'port' => 6379,
    'timeout' => 0,
    'dbindex' => 0,
  ),

  'trashbin_retention_obligation' => '7,30',
  'versions_retention_obligation' => '15,90',
);
```

• Storage type and setup
```

NFS->
10.1.10.102:/mnt/md1/public/owncloud  /mnt/owncloud nfs soft,intr,rsize=8192,wsize=8192
```

### PR DESCRIPTION
**Pull request by VicDeo**
…ename. Fixes #22450

### CODE REVIEW

File: apps/files_versions/lib/Storage.php [hunk 1]
```diff
@@ -532,13 +532,15 @@ private static function getAllVersions($uid) {
...
 
 			foreach ($files as $file) {
+				$fileData = $file->getData();
+				$filePath = $dir . '/' . $fileData['name'];
 				if ($file['type'] === 'dir') {
-					array_push($dirs, $file['path']);
+					array_push($dirs, $filePath);
 				} else {
-					$versionsBegin = strrpos($file['path'], '.v');
+					$versionsBegin = strrpos($$filePath, '.v');
```
Comment (PVince81) [inline apps/files_versions/lib/Storage.php]: two dollars? looks suspicious

Mind adjusting the unit test as well to properly cover this, if possible ?

Comment (VicDeo) [inline apps/files_versions/lib/Storage.php]: @PVince81 sorry, it slipped through. Surely only one.
I don't know how to reproduce it.
 The only thing I know that it works and that getDirectoryContents produces path relative to datadir instead if view root under some unclear conditions.

**Comment by PVince81** [comment]
Weird, so this means we can't trust the output of `getDirectoryContents` ?
There might be other places where shit can go wrong because of that :frowning: 

Would be good to find out what the root cause is.

**Comment by PVince81** [comment]
@VicDeo can you add a unit test ?

**Comment by icewind1991** [comment]
You can't trust the output of `$fileInfo['path']`

`$fileInfo->getPath()` can be trusted though.

The difference between the two is due to legacy reasons and I'm not sure if I can fix `['path']` without breaking something

**Comment by PVince81** [comment]
@icewind1991 @VicDeo so should we merge this without unit tests ??

**Comment by VicDeo** [comment]
@PVince81 `getAllVersions` is private and static so I'd say yes.

**Comment by PVince81** [comment]
Ok then...

@icewind1991 can you confirm that this PR is safe ?

**Comment by icewind1991** [comment]
:+1: looks good

**Comment by PVince81** [comment]
Ok, thanks guys. I rebased one last time for CI.

:+1:

**Comment by VicDeo** [comment]
@cmonteroluque initailly reported for 8.2. Backport?

**Comment by PVince81** [comment]
@VicDeo okay please backport down to 8.2

CC @dragotin @DeepDiver1975

**Comment by VicDeo** [comment]
Stable9: https://github.com/owncloud/core/pull/25039
Stable8.2: https://github.com/owncloud/core/pull/25040

**Pull request merged**